File: 0152-maximum-product-subarray/0152-maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        // if(nums.size()==1)return nums[0];
        // int maxi=0;
        // for(int i=0;i<nums.size();i++){
        //     int prod=1;
        //     for(int j=i;j<nums.size();j++){
        //         prod=prod*nums[j];
        //         maxi=max(maxi,prod);
        //     }
        // }
        // return maxi;
        //OPtimal solution 
        if(nums.size()==1)return nums[0];
        int pre =1,suff=1;
        int maxi=0;
        for(int i=0;i<nums.size();i++){
            if(pre==0) pre=1;
            if(suff==0)suff=1;
            pre=pre*nums[i];
            suff=suff*nums[nums.size()-i-1];
            maxi=max(maxi,max(pre,suff));
        } 
        return maxi;
    }
};

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int max=nums[0];
        int min=nums[0];
        int ans=nums[0];
        for (int i=1;i<nums.length;i++){
            int curr= nums[i];
            int tempMax = Math.max(curr,Math.max(max*curr,min*curr));
            int tempmin=Math.min(curr,Math.min(max*curr,min*curr));
            max = tempMax;
            min=tempmin;

            ans=Math.max(ans,max);
        }
        return ans;

    }
}